New Component: HoverCard

atneik opened the pull request:
#### Description of changes

Adding a new component: HoverCard

#### Focus areas to test

- Card's Expanding behavior.
- Perf
- Customization 
- Changes to Callout. Make sure no regressions.

https://odsp-int.azurewebsites.net/anihan/hovercard/#/examples/hovercard

dzearing commented:
no screenshots?

jakob101 commented:
Heya,

You seem to have an issue with scrollbar, I'm using Chrome on Win:
![Issue with scrollbar](https://media.giphy.com/media/l4FGn4dyHMNb46QKs/giphy.gif)

jakob101 commented on packages/office-ui-fabric-react/src/components/Callout/CalloutContent.tsx:
```diff
@@ -336,4 +341,25 @@ export class CalloutContent extends BaseComponent<ICalloutProps, ICalloutState>
+  private _setHeightOffsetEveryFrame(): void {
+    if (this._calloutElement && this.props.finalHeight) {
+      this._setHeightOffsetTimer = this._async.requestAnimationFrame(() => {
+        const calloutMainElem = this._calloutElement.firstChild as HTMLElement;
```
Dunno how expensive this is (probably not at all), but it could be moved outside the requestAnimationFrame.

jakob101 commented on packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.Props.ts:
```diff
@@ -0,0 +1,100 @@
+/**
+ * HoverCard component props.
+ */
+export interface IHoverCardProps extends React.HTMLAttributes<HTMLDivElement | HoverCard> {
```
Can we control stickiness too (to not disappear even when you mouse out)? We'd love to use this and we need that support, as our LPC is sticky in SPhome and ODSP currently.

mtennoe replied:
+1, a stickiness bool prop would be nice to have

jakob101 commented on packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.styles.ts:
```diff
@@ -0,0 +1,45 @@
+        height: '0',
+        'overflow-y': 'auto',
+        'pointer-events': 'auto',
+        transition: 'height 0.467s cubic-bezier(0.5, 0, 0, 1)'
```
help me understand why you need this?

mtennoe commented on packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.tsx:
```diff
@@ -0,0 +1,131 @@
+import { getStyles } from './HoverCard.styles';
+
+export enum HoverCardMode {
+  condensed,
```
Is "compact" a more descriptive name?

mtennoe commented:
Awesome work!

dzearing commented:
@jakob101  @mtennoe Let us know when you're approved, we'll get it merged.

atneik commented:
@jakob101 @mtennoe Thanks for your comments. I addressed the scroll issue (I was able to repro it only on Edge), added the Sticky boolean and other fixes. Kindly give it a look again.

micahgodbolt commented on packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.tsx:
```diff
@@ -0,0 +1,161 @@
+import { getStyles } from './HoverCard.styles';
+
+export enum HoverCardMode {
+  compact,
```
use explicit enum syntax

compact = 0,
expanded = 1

micahgodbolt commented on packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.tsx:
```diff
@@ -0,0 +1,161 @@
+  public componentWillMount() {
+    this._async.setTimeout(() => {
+      this.setState({
+        mode: HoverCardMode.expanded
```
should this be wrapped in an "if (!props.openExpanded)"

micahgodbolt commented on packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.tsx:
```diff
@@ -0,0 +1,161 @@
+
+    return (
+      <Callout
+        componentRef={ c => this._callout = c }
```
use _resolveRef instead?

micahgodbolt commented on packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.tsx:
```diff
@@ -0,0 +1,161 @@
+        minPagePadding={ 24 }
+      >
+        <div
+          { ...{ onFocusCapture: this.props.onEnter } }
```
{ ...{ onFocusCapture: this.props.onEnter } } === { onFocusCapture: this.props.onEnter }

Do we need a spread here?

micahgodbolt commented on packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.tsx:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  @autobind
+  private _onRenderCompactContent(): JSX.Element {
```
Typically when onRenderCompactContent is a function and onRenderCompactContent exists in props, the prop is able to replace the internal function. Can we use another name here? We're actually rendering the compact card here, right? maybe onRenderCompactCard, then inside is the CompactContend function.

micahgodbolt commented on packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.tsx:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  @autobind
+  private _onRenderExpandedContent(): JSX.Element {
```
Same as above.

micahgodbolt commented on packages/office-ui-fabric-react/src/components/HoverCard/HoverCardHost.tsx:
```diff
@@ -0,0 +1,126 @@
+      <div
+        className={ css(this._styles.host) }
+        ref={ this._resolveRef('_hoverCardHost') }
+        { ...{ onFocusCapture: this._cardOpen } }
```
same spread operator question. can these be removed?

atneik commented:
@micahgodbolt Addressed what we talked about :)

micahgodbolt commented on packages/office-ui-fabric-react/src/components/HoverCard/ExpandingCard.Props.ts:
```diff
@@ -0,0 +1,106 @@
+  targetElement?: HTMLElement;
+
+  /**
+   * Callback when the ExpandingCard tries to open
```
should this be expand instead of open?

micahgodbolt replied:
also, onEnter/onDismiss aren't very descriptive. onExpand onCollapse would be better.

micahgodbolt replied:
actually...last idea here. since we currently have an enum for the card modes, how about simply have onModeChange, and pass in old/new mode. This will support more modes in the future without creating even more callbacks.

atneik replied:
Updated the name and comments for both the callbacks

micahgodbolt commented on packages/office-ui-fabric-react/src/components/HoverCard/ExpandingCard.Props.ts:
```diff
@@ -0,0 +1,106 @@
+  onEnter?: (ev?: any) => void;
+
+  /**
+   * Callback when the ExpandingCard tries to close
```
collapse vs close?

micahgodbolt commented on packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.Props.ts:
```diff
@@ -0,0 +1,72 @@
+   * Enables instant open of the full card upon click
+   * @default false
+   */
+  instantOpenOnClick?: boolean;
```
does this replace the hover behavior or just add to it?

atneik replied:
Just add to it
